vcstool/clients/vcs_base.py

```diff
@@ -115,20 +132,145 @@ def test_url(url, retry=2, retry_period=1, timeout=10):
+        opener = build_opener(auth_handler)
+        return _urlopen_retry(retry, retry_period)(opener.open)(
+            request, timeout=timeout)
+
+    # If we have only a password, use token auth
+    elif password:
+        request.add_header('PRIVATE-TOKEN', password)
+        return _urlopen_retry(retry, retry_period)(urlopen)(
+            request, timeout=timeout)
+
+    return None
+
+
+def _credentials_for_machine(machine):
+    # First check the default .netrc file, if any-- it takes precedence over
+    # everything else
+    credentials = _credentials_for_machine_in_file(None, machine)
+    if credentials:
+        return credentials
+
+    # If that file either didn't exist or didn't match for the machine,
+    # check the user auth directory for vcstool
+    credentials = _credentials_for_machine_in_dir(
+        appdirs.user_config_dir(_APPDIRS_PROJECT_NAME), machine)
+    if credentials:
+        return credentials
+
+    # Finally, check the system-wide auth directories for vcstool, in order
+    auth_path = appdirs.site_config_dir(_APPDIRS_PROJECT_NAME, multipath=True)
+    for site_config_dir in auth_path.split(':'):
+        credentials = _credentials_for_machine_in_dir(
+            site_config_dir, machine)
+        if credentials:
+            return credentials
+
+    return None
+
+
+def _credentials_for_machine_in_dir(directory, machine):
+    # The idea here is similar to how Debian handles authenticated apt repos:
+    # https://manpages.debian.org/testing/apt/apt_auth.conf.5.en.html
+
+    # Check the auth.conf file first
+    auth_file_path = os.path.join(
+        directory, _AUTHENTICATION_CONFIGURATION_FILE)
+    credentials = _credentials_for_machine_in_file(auth_file_path, machine)
+    if credentials:
+        return credentials
+
+    # If that file either didn't exist or didn't match for the machine, check
+    # the .conf files in the configuration parts dir
+    configuration_parts_dir = os.path.join(
+        directory, _AUTHENTICATION_CONFIGURATION_PARTS_DIR)
+    auth_files = glob.glob(os.path.join(configuration_parts_dir, '*.conf'))
+    for auth_file in sorted(auth_files):
+        auth_file_path = os.path.join(configuration_parts_dir, auth_file)
+        credentials = _credentials_for_machine_in_file(auth_file_path, machine)
+        if credentials:
+            return credentials
+
+    # Nothing matched
+    return None
+
+
+def _credentials_for_machine_in_file(filename, machine):
+    credentials = None
+    try:
+        credentials = netrc.netrc(file=filename).authenticators(machine)
+    except EnvironmentError as e:
+        # Don't error just because the file didn't exist or we didn't have
+        # permission to access it. Catching this situation this way to be
+        # compatible with python 2 and 3.
+        if e.errno not in (errno.ENOENT, errno.EACCES):
+            raise
+    except netrc.NetrcParseError:
+        # If this file had issues, don't error out so we can try fallbacks
```

Comment:
Would a warning be more helpful to the user rather then silently ignoring it? On the other hand this behavior might match how other programs using `.netrc` work.

Reply:
We're a little bit in uncharted territory here, to be honest. My model for the design is `apt`, but one generally runs that as root, so running into permissions issues accessing these files isn't normally an issue. It could be here, if the machine admin only wanted members of a particular group to be able to fetch repositories using these credentials. I had mixed feelings about warning about this or quietly proceeding, so I'm happy adding a warning if you like.